common/state/state_kv.cpp

```diff
@@ -153,42 +153,158 @@ void pstate::cache_slots::release(data_node** dn)
     {
         dn_queue_.push(*dn);
     }
-    catch (std::exception& e)
+    catch (const std::exception& e)
     {
-        SAFE_LOG_EXCEPTION("cache_slots() release");
+        SAFE_LOG_EXCEPTION("cache_slots release error");
         throw;
     }
     // delete original pointer
     *dn = NULL;
 }
 
-pstate::block_offset_t* pstate::trie_node::goto_next_offset(trie_node_header_t* header)
+bool pstate::free_space_collector::are_adjacent(const block_offset_t& bo1, const unsigned& length1, const block_offset_t& bo2)
+{
+    pdo::error::ThrowIf<pdo::error::RuntimeError>(bo1.block_num > bo2.block_num ||
+        (bo1.block_num == bo2.block_num && bo1.bytes > bo2.bytes),
+        "free space collector, adjancency error");
+    block_offset_t bo = bo1;
+    data_node::advance_block_offset(bo, length1);
+    return (bo == bo2);
+}
+
+void pstate::free_space_collector::collect(const block_offset_t& bo, const unsigned int& length)
 {
-    if (header->hasNext)
+    if(length ==0) // nothing to collect
     {
-        trie_node_h_with_n_t* p = (trie_node_h_with_n_t*)header;
-        return &(p->next_offset);
+        return;
+    }
+
+    for(auto it = free_space_collection.begin(); ; it++)
+    {
+        if(it == free_space_collection.end() || //if end of vector is reached
+            (bo.block_num < it->bo.block_num || //or the item location preceeds the current one, insert!
+            (bo.block_num == it->bo.block_num && bo.bytes < it->bo.bytes)))
+        {
+            //the bo should be placed at this point
+            free_space_item_t fsi = {bo, length};
+
+            //first check if it can merge with previous
+            if(it != free_space_collection.begin())
+            {
+                auto prev_it = std::prev(it);
+                if(are_adjacent(prev_it->bo, prev_it->length, fsi.bo))
+                {
+                    //update item to be inserted
+                    fsi.bo = prev_it->bo;
+                    fsi.length += prev_it->length;
+                    //remove previous item
+                    it = free_space_collection.erase(prev_it);
+                }
+            }
+            //also, check if it can merge with current
+            if(it != free_space_collection.end() && are_adjacent(fsi.bo, fsi.length, it->bo))
+            {
+                //item to be inserted is the same, just increase length
+                fsi.length += it->length;
+                it = free_space_collection.erase(it);
+            }
+            //any merge done, now insert
+            free_space_collection.insert(it, fsi);
+            break;
+        }
+
+        //no insert, go to next
```

Comment:
why not just put the end() check in the for loop so you know you'll stop?

Reply:
changed it

anyway, the previous version DID stop: if it == free_space_collection.end() ... then break;